src/miner.erl

```diff
@@ -600,9 +445,232 @@ terminate(Reason, _State) ->
+    SortedTransactions =
+        lists:sort(fun blockchain_txn:sort/2, [T || T <- Txns, not txn_is_rewards(T)]),
+    {ValidTransactions, InvalidTransactions0} = blockchain_txn:validate(SortedTransactions, Chain),
+    %% InvalidTransactions0 is a list of tuples in the format {Txn, InvalidReason}
+    %% we dont need the invalid reason here so need to remove the tuple format
+    %% and have a regular list of txn items in prep for returning to hbbft.
+    InvalidTransactions = [InvTxn || {InvTxn, _InvalidReason} <- InvalidTransactions0],
+    Ledger = blockchain:ledger(Chain),
+    case blockchain_election:has_new_group(ValidTransactions) of
+            {true, _, ConsensusGroupTxn, _} ->
+                Epoch = ElectionEpoch0 + 1,
+                Start = EpochStart0 + 1,
+                End = BlockHeightCurr,
+                RewardsMod =
+                    case blockchain:config(?rewards_txn_version, Ledger) of
+                         {ok, 2} -> blockchain_txn_rewards_v2;
+                         _       -> blockchain_txn_rewards_v1
+                     end,
+                {ok, Rewards} = RewardsMod:calculate_rewards(Start, End, Chain),
+                lager:debug("RewardsMod: ~p, Rewards: ~p~n", [RewardsMod, Rewards]),
+                RewardsTxn = RewardsMod:new(Start, End, Rewards),
+                %% To cut down on the size of group txn blocks, which we'll
+                %% need to fetch and store all of to validate snapshots, we
+                %% discard all other txns for this block.
+                Transactions =
+                    lists:sort(
+                        %% TODO Rename blockchain_txn:sort to blockchain_txn:(compare|cmp)
+                        fun blockchain_txn:sort/2,
+                        [RewardsTxn, ConsensusGroupTxn]
+                     ),
+                {Epoch, BlockHeightNext, Transactions, InvalidTransactions};
+            _ ->
+                {ElectionEpoch0, EpochStart0, ValidTransactions, InvalidTransactions}
+        end.
+
+-spec txn_is_rewards(blockchain_txn:txn()) -> boolean().
+txn_is_rewards(Txn) ->
+    Rewards = [blockchain_txn_rewards_v1, blockchain_txn_rewards_v2],
+    lists:member(blockchain_txn:type(Txn), Rewards).
+
+-spec metadata_to_seen_bbas(metadata()) ->
+    [{{pos_integer(), binary()}, binary()}].
+metadata_to_seen_bbas(M) ->
+    [{{J, S}, B} || {J, #{seen := S, bba_completion := B}} <- metadata_only_v2(M)].
+
+-spec metadata_only_v2(metadata()) ->
+    [{non_neg_integer(), metadata_v2()}].
+metadata_only_v2(Metadata) ->
+    lists:filter(fun ({_, M}) -> is_map(M) end, Metadata).
+
+-spec meta_to_stamp_hashes(metadata()) ->
+    {
+        Stamps :: [integer()],
+        Hashes :: [blockchain_block:hash()]
+    }.
+meta_to_stamp_hashes(Metadata) ->
+    lists:unzip([metadata_as_v1(M) || {_, M} <- Metadata]).
+
+-spec metadata_as_v1(metadata_v1() | metadata_v2()) -> metadata_v1().
+metadata_as_v1(#{head_hash := H, timestamp := S}) -> {S, H}; %% v2 -> v1
```

Comment:
honestly at this point we can remove all the v1 code.  all that it does is complicate things, and nothing will generate it anymore.